tools/codegen/gen_functionalization_type.py

```diff
@@ -0,0 +1,426 @@
+    else:
+        return cpp.name(func)
+
+def is_tensor_like(a: Union[Argument, TensorOptionsArguments, SelfArgument]) -> bool:
+    return isinstance(a, SelfArgument) or (isinstance(a, Argument) and a.type.is_tensor_like())
+
+# unwraps all tensor-like arguments, returning:
+# (1) a string containing all of the logic that does the unwrapping
+# (2) a context, to be used by translate(), with all of the relevant bindings.
+def unwrap_tensor_args(sig: DispatcherSignature) -> Tuple[str, List[Binding]]:
+    context: List[Binding] = []
+    unwrapped_tensor_args: List[str] = []
+    for arg in sig.arguments():
+        if is_tensor_like(arg.argument):
+            # for tensor inputs, we want to unwrap them before passing them into the redispatch calls.
+            unwrapped_name = f'{arg.name}_'
+            unwrapped_tensor_args.append(
+                f'auto {unwrapped_name} = at::functionalization::impl::from_functional_tensor({arg.name});')
+            context.append(arg.with_name(unwrapped_name))
+        else:
+            # for non-tensor inputs, we want to pass them directly into the redispatch calls.
+            context.append(arg)
+    unwrap_tensor_args_str = '\n      '.join(unwrapped_tensor_args)
+    return unwrap_tensor_args_str, context
+
+# converts  all tensor-like arguments to meta tensors, which are used to compute stride info. Returns:
+# (1) a string containing all of the logic that does the conversions.
+# (2) a context, to be used by translate(), with all of the relevant bindings.
+def convert_to_meta_tensors(sig: DispatcherSignature) -> Tuple[str, List[Binding]]:
+    context: List[Binding] = []
+    unwrapped_tensor_args: List[str] = []
+    for arg in sig.arguments():
+        if is_tensor_like(arg.argument):
+            # for tensor inputs, we want to unwrap them before passing them into the redispatch calls.
+            # for tensor inputs, we want to unwrap them before passing them into the redispatch calls.
+            a_ = arg.name
+            unwrapped_name = f'{arg.name}_meta'
+            unwrapped_tensor_args.append(
+                f"auto {unwrapped_name} = at::native::empty_strided_meta({a_}.sizes(), {a_}.strides(), \
+/*dtype=*/c10::make_optional({a_}.scalar_type()), /*layout=*/c10::make_optional({a_}.layout()), \
+/*device=*/c10::make_optional(c10::Device(kMeta)), /*pin_memory=*/c10::nullopt);"
+            )
+            context.append(arg.with_name(unwrapped_name))
+        else:
+            # for non-tensor inputs, we want to pass them directly into the redispatch calls.
+            context.append(arg)
+    unwrap_tensor_args_str = '\n      '.join(unwrapped_tensor_args)
+    return unwrap_tensor_args_str, context
+
+# The function lambdas generated for each view op in the functionalization pass are of the form
+# [captures](input_args) -> return_type {
+#     return name(call_args);
+# }
+# This function returns the tuple of (input_args, name, return_type, call_args)
```

Comment:
Probably should just define a miniature data class you can refer to the elements by name. Esp helpful because input_args and call_args have the same type so typechecker can't help you if you mix them up